### PR TITLE
Adding Trimming tests for StackFrameHelper and AppDomain

### PR DESCRIPTION
**Pull request by joperezr**
Contributes to https://github.com/dotnet/runtime/issues/37677

cc: @layomia @eerhardt 

These two tests are the ones needed to cover the first half of the list on @layomia's issue. Once this is merged I'll check the corresponding checkboxes on that.

### CODE REVIEW

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+using System.Security.Principal;
+using System.Threading;
+
+class Program
+{
+    static int Main(string[] args)
+    {
+        // Thread.CurrentPrincipal calls AppDomain.CurrentDomain.GetThreadPrincipal() which
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs:8]: This path is taken only when you set PrincipalPolicy to `PrincipalPolicy.WindowsPrincipal` or `PrincipalPolicy.UnauthenticatedPrincipal`. As written, this test won't exercise these codepaths.

File: src/libraries/System.Runtime/tests/TrimmingTests/StackFrameHelperTest.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Diagnostics;
+
+class Program
+{
+    static int Main(string[] args)
+    {
+        // StackTrace constructor calls CaptureStackTrace method which in turn will call
+        // StackFrameHelper.InitializeSourceInfo which has annotations as it will require
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/TrimmingTests/StackFrameHelperTest.cs:9]: `InitializeSourceInfo` has a big try/catch and ignores any exceptions. You also need to validate the result to tell whether `InitializeSourceInfo` worked properly.

File: src/libraries/System.Runtime/tests/TrimmingTests/StackFrameHelperTest.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        // StackTraceSymbols.ctor and StackTraceSymbols.GetSourceLineInfo
+        StackTrace thisTrace = new StackTrace(true);
```
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/StackFrameHelperTest.cs:10]: (nit) can you name this parameter?  `new StackTrace(fNeedFileInfo: true);`

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        // contains annotation attributes and will require WindowsPrincipal.GetDefaultInstance
+        // and GenericPrincipal.GetDefaultInstance
+        AppDomain.CurrentDomain.SetPrincipalPolicy(PrincipalPolicy.WindowsPrincipal);
```
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs:11]: #36864 had 2 policies: Windows and Generic. We should test both.

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        IPrincipal principal = Thread.CurrentPrincipal;
```
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs:12]: Should we verify this is a `WindowsPrincipal`?

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs (hunk 4 above)
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs:11]: I assume this will only work on a Windows machine, not OSX or Linux.

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs (hunk 5 above)
Comment (joperezr) [reply src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs]: No need, if the method get's trimmed then the re will be an uncaught exception at runtime for the missing type/member which will return a failing exit code.

File: src/libraries/System.Runtime/tests/TrimmingTests/System.Runtime.TrimmingTests.proj [hunk 6]
```diff
@@ -1,5 +1,20 @@
...
   <Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Build.props))" />
 
+  <ItemGroup>
+    <TestConsoleAppSourceFiles Include="AppDomainGetThreadGenericPrincipalTest.cs" />
+    <TestConsoleAppSourceFiles Include="AppDomainGetThreadWindowsPrincipalTest.cs">
+      <SkipOnTestRuntimes>osx-x64;linux-x64</SkipOnTestRuntimes>
+    </TestConsoleAppSourceFiles>
+    <TestConsoleAppSourceFiles Include="GenericArraySortHelperTest.cs" />
+    <TestConsoleAppSourceFiles Include="StackFrameHelperTest.cs">
+      <!-- There is a bug with the linker where it is corrupting the pdbs while trimming
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/TrimmingTests/System.Runtime.TrimmingTests.proj:11]: Add link to the issue tracking this bug?

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadGenericPrincipalTest.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Security.Principal;
+using System.Threading;
+using System.Diagnostics;
+
+class Program
+{
+    static int Main(string[] args)
+    {
+    	// Thread.CurrentPrincipal calls AppDomain.CurrentDomain.GetThreadPrincipal() which
```
Comment (eerhardt) [inline src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadGenericPrincipalTest.cs:10]: (nit) whitespace is mis-aligned here and in AppDomainGetThreadWindowsPrincipalTest.cs

File: src/libraries/restore/runtime/runtime.depproj [hunk 8]
```diff
@@ -69,7 +69,6 @@
       <BinPlaceItem Include="@(MonoIncludeFiles)">
         <DestinationSubDirectory>include/%(RecursiveDir)</DestinationSubDirectory>
       </BinPlaceItem>
-      <BinPlaceItem Remove="@(BinPlaceItem)" Condition="'$(RuntimeFlavor)' != 'Mono' and '%(BinPlaceItem.Filename)' == 'System.Private.CoreLib'" />
```
Comment (eerhardt) [inline src/libraries/restore/runtime/runtime.depproj:72]: Is this change intended? Why is it necessary?

File: src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs (hunk 5 above)
Comment (eerhardt) [reply src/libraries/System.Runtime/tests/TrimmingTests/AppDomainGetThreadPrincipalTest.cs]: This could very easily change to return `null` in the future if an exception occurs, and this test wouldn't catch it. I would at least call `principal.GetType()` and check the name is correct.

File: src/libraries/restore/runtime/runtime.depproj (hunk 8 above)
Comment (joperezr) [reply src/libraries/restore/runtime/runtime.depproj:72]: I added it back when I merged the Trimming infrastructure because of how we used to binplace the runtime packs but it was actually not required so I'm just cleaning that up now.

**Review by eerhardt — APPROVED** [review]
Looks good to me.
